dvc/remote/local.py

```diff
@@ -348,48 +428,121 @@ def _process(
         if jobs is None:
             jobs = remote.JOBS
 
-        status_info = self.status(
-            named_cache,
+        dir_status, file_status, dir_paths = self._status(
+            named_caches,
             remote,
             jobs=jobs,
             show_checksums=show_checksums,
             download=download,
+            drop_index=drop_index,
         )
 
-        plans = self._get_plans(download, remote, status_info, status)
+        dir_plans = self._get_plans(download, remote, dir_status, status)
+        file_plans = self._get_plans(download, remote, file_status, status)
 
-        if len(plans[0]) == 0:
+        if len(dir_plans[0]) + len(file_plans[0]) == 0:
             return 0
 
-        if jobs > 1:
-            with ThreadPoolExecutor(max_workers=jobs) as executor:
-                fails = sum(executor.map(func, *plans))
-        else:
-            fails = sum(map(func, *plans))
+        with ThreadPoolExecutor(max_workers=jobs) as executor:
+            if download:
+                fails = sum(executor.map(func, *dir_plans))
+                fails += sum(executor.map(func, *file_plans))
+            else:
+                # for uploads, push files first, and any .dir files last
+
+                file_futures = {}
+                for from_info, to_info, name in zip(*file_plans):
+                    file_futures[to_info] = executor.submit(
+                        func, from_info, to_info, name
+                    )
+                dir_futures = {}
+                for from_info, to_info, name in zip(*dir_plans):
+                    wait_futures = {
+                        future
+                        for file_path, future in file_futures.items()
+                        if file_path in dir_paths[to_info]
+                    }
+                    dir_futures[to_info] = executor.submit(
+                        self._dir_upload,
+                        func,
+                        wait_futures,
+                        from_info,
+                        to_info,
+                        name,
+                    )
+                fails = sum(
+                    future.result()
+                    for future in concat(
+                        file_futures.values(), dir_futures.values()
+                    )
+                )
 
         if fails:
             if download:
+                remote.index.invalidate()
                 raise DownloadError(fails)
             raise UploadError(fails)
+        elif not download:
+            pushed_checksums = list(
+                map(self.path_to_checksum, dir_plans[0] + file_plans[0])
+            )
+            logger.debug(
+                "Adding {} pushed checksums to index".format(
+                    len(pushed_checksums)
+                )
+            )
+            remote.index.update(pushed_checksums)
+            remote.index.save()
```

Comment:
make sure that it is thread/multi-process safe (we do allow multiple DVC commands in parallel) ... check other this new index functions if they are thread safe?